pkg/login/login.go

```diff
@@ -0,0 +1,83 @@
+package login
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/dcos/dcos-cli/pkg/httpclient"
+	"github.com/sirupsen/logrus"
+)
+
+// Client is able to detect available login providers and login to DC/OS.
+type Client struct {
+	http   *httpclient.Client
+	logger *logrus.Logger
+}
+
+// NewClient creates a new login client.
+func NewClient(baseClient *httpclient.Client, logger *logrus.Logger) *Client {
+	return &Client{
+		http:   baseClient,
+		logger: logger,
+	}
+}
+
+// Providers returns the supported login providers for a given DC/OS cluster.
+func (c *Client) Providers() (map[string]*Provider, error) {
+	resp, err := c.http.Get("/acs/api/v1/auth/providers")
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	providers := make(map[string]*Provider)
+	if resp.StatusCode == 200 {
+		err := json.NewDecoder(resp.Body).Decode(&providers)
+		if err == nil {
+			return providers, nil
+		}
+	}
+
+	c.logger.Info("Falling back to the WWW-Authenticate challenge.")
+
+	// This is for DC/OS EE 1.7/1.8 and DC/OS Open Source.
+	authHeader, err := c.challengeAuth()
+	if err != nil {
+		return nil, err
+	}
+
+	switch authHeader {
+	case "oauthjwt":
+		// DC/OS Open Source
+		providers["dcos-oidc"] = defaultOIDCImplicitFlowProvider()
+	case "acsjwt":
+		// DC/OS EE 1.7/1.8
+		providers["dcos-users"] = defaultDCOSUIDPasswordProvider()
+	default:
+		return nil, fmt.Errorf("unsupported WWW-Authenticate challenge '%s'", authHeader)
+	}
+	return providers, nil
+}
+
+// challengeAuth sends an unauthenticated HTTP request to a DC/OS well-known resource.
```

Comment:
Why `challenge` in the function name?

Reply:
From https://tools.ietf.org/html/rfc2617 : 
> HTTP provides a simple challenge-response authentication mechanism that MAY be used by a server to challenge a client request and by a client to provide authentication information.

The current one in Python also has the `challenge` semantic : `header_challenge_auth`.